### PR TITLE
[Security] Lazy load user providers

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

**Comment by chalasr** [comment]
build failure with high deps is normal

**Review by ogizanagi — APPROVED** [review]

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Component/Security/Core/User/ChainUserProvider.php [hunk 1]
```diff
@@ -26,7 +26,10 @@ class ChainUserProvider implements UserProviderInterface
 {
     private $providers;
 
-    public function __construct(array $providers)
+    /**
+     * @param iterable $providers
+     */
+    public function __construct(/* iterable */ $providers)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/User/ChainUserProvider.php]: the comment should be removed

File: src/Symfony/Component/Security/Http/Firewall/ContextListener.php [hunk 2]
```diff
@@ -44,7 +44,7 @@ class ContextListener implements ListenerInterface
     private $registered;
     private $trustResolver;
 
-    public function __construct(TokenStorageInterface $tokenStorage, array $userProviders, $contextKey, LoggerInterface $logger = null, EventDispatcherInterface $dispatcher = null, AuthenticationTrustResolverInterface $trustResolver = null)
+    public function __construct(TokenStorageInterface $tokenStorage, /* iterable */ $userProviders, $contextKey, LoggerInterface $logger = null, EventDispatcherInterface $dispatcher = null, AuthenticationTrustResolverInterface $trustResolver = null)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/ContextListener.php]: comment to be removed, that's the job of a docblock, we should not follow this practice IMHO

**Review by nicolas-grekas — APPROVED** [review]
with minor comments

**Comment by chalasr** [comment]
@nicolas-grekas comments addressed. build failure unrelated (see #23360)

File: src/Symfony/Component/Security/Http/Tests/Firewall/ContextListenerTest.php [hunk 3]
```diff
@@ -47,19 +47,6 @@ public function testItRequiresContextKey()
         );
     }
 
-    /**
-     * @expectedException \InvalidArgumentException
-     * @expectedExceptionMessage User provider "stdClass" must implement "Symfony\Component\Security\Core\User\UserProviderInterface
```
Comment (stof) [inline src/Symfony/Component/Security/Http/Tests/Firewall/ContextListenerTest.php]: we should still have a test covering this case. You have not removed the exception entirely. You just moved it later in the usage.

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Security/Http/Tests/Firewall/ContextListenerTest.php (hunk 3 above)
Comment (chalasr) [inline src/Symfony/Component/Security/Http/Tests/Firewall/ContextListenerTest.php]: test readded

**Comment by chalasr** [comment]
ready.

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Component/Security/Http/Firewall/ContextListener.php [hunk 4]
```diff
@@ -44,18 +44,20 @@ class ContextListener implements ListenerInterface
...
-    public function __construct(TokenStorageInterface $tokenStorage, array $userProviders, $contextKey, LoggerInterface $logger = null, EventDispatcherInterface $dispatcher = null, AuthenticationTrustResolverInterface $trustResolver = null)
+    /**
+     * @param TokenStorageInterface                     $tokenStorage
+     * @param iterable|UserProviderInterface[]          $userProviders
+     * @param string                                    $contextKey
+     * @param LoggerInterface|null                      $logger
+     * @param EventDispatcherInterface|null             $dispatcher
+     * @param AuthenticationTrustResolverInterface|null $trustResolver
+     */
+    public function __construct(TokenStorageInterface $tokenStorage, $userProviders, $contextKey, LoggerInterface $logger = null, EventDispatcherInterface $dispatcher = null, AuthenticationTrustResolverInterface $trustResolver = null)
```
Comment (chalasr) [inline src/Symfony/Component/Security/Http/Firewall/ContextListener.php]: full docblock to be removed + `iterable` typehint to be added to the `$userProviders` arg when merging into master.

File: src/Symfony/Component/Security/Core/User/ChainUserProvider.php [hunk 5]
```diff
@@ -26,7 +26,10 @@ class ChainUserProvider implements UserProviderInterface
 {
     private $providers;
 
-    public function __construct(array $providers)
+    /**
+     * @param iterable|UserProviderInterface[] $providers
+     */
+    public function __construct($providers)
```
Comment (chalasr) [inline src/Symfony/Component/Security/Core/User/ChainUserProvider.php]: same, docblock to typehint when merging into master

**Comment by nicolas-grekas** [comment]
Thank you @chalasr.

**Pull request merged**